File: LeetCode/0049-group-anagrams/0049-group-anagrams.java
import java.util.Optional;

class Solution {
    public List<List<String>> groupAnagrams(String[] strs) {
        Map<String, List<String>> anagramGrps = new HashMap<>();

        for (String str : strs) {
            char[] word = str.toCharArray();

            Arrays.sort(word);

            String key = new String(word);

            anagramGrps.putIfAbsent(key, new ArrayList<>());
            anagramGrps.get(key).add(str);
        }

        return new ArrayList<>(anagramGrps.values());
    }
}

File: LeetCode/0049-group-anagrams/0049-group-anagrams.kt
class Solution {
    fun groupAnagrams(strs: Array<String>): List<List<String>> {
      val anagramGrps: MutableMap<String, MutableList<String>> = mutableMapOf()
      for (str in strs) {
        val key = str.toCharArray().sortedArray().joinToString("")
        anagramGrps.getOrPut(key) { mutableListOf() }.add(str)
      }
      return ArrayList(anagramGrps.values)
    }
}